# Trabajo 1 - Nicolas Maldonado - AX840282
## Preparacion de Datos
8 / Diciembre / 2023<br>
Master en Ciencia de Datos y Aprendizaje Automatico

### Descripción: Passenger Safisfaccion

### Objetivo:
Que factores estan altamente correlacionados con la satisfaccion de los usuarios al volar en una aerolinea.

Empezamos cargando las librerias y el dataset, https://www.kaggle.com/datasets/teejmahal20/airline-passenger-satisfaction?select=test.csv

In [1]:
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
import sklearn

import pandas as pd
import numpy as np

In [2]:
df =  pd.read_csv('Datasets/Airplane Satisfaction/train.csv')
df

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0                0   70172    Male     Loyal Customer   13  Personal Travel   
1                1    5047    Male  disloyal Customer   25  Business travel   
2                2  110028  Female     Loyal Customer   26  Business travel   
3                3   24026  Female     Loyal Customer   25  Business travel   
4                4  119299    Male     Loyal Customer   61  Business travel   
...            ...     ...     ...                ...  ...              ...   
103899      103899   94171  Female  disloyal Customer   23  Business travel   
103900      103900   73097    Male     Loyal Customer   49  Business travel   
103901      103901   68825    Male  disloyal Customer   30  Business travel   
103902      103902   54173  Female  disloyal Customer   22  Business travel   
103903      103903   62567    Male     Loyal Customer   27  Business travel   

           Class  Flight Distance  Inflight wifi service  \
0       Eco Plus              460                      3   
1       Business              235                      3   
2       Business             1142                      2   
3       Business              562                      2   
4       Business              214                      3   
...          ...              ...                    ...   
103899       Eco              192                      2   
103900  Business             2347                      4   
103901  Business             1995                      1   
103902       Eco             1000                      1   
103903  Business             1723                      1   

        Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                       4  ...                       5   
1                                       2  ...                       1   
2                                       2  ...                       5   
3                                       5  ...                       2   
4                                       3  ...                       3   
...                                   ...  ...                     ...   
103899                                  1  ...                       2   
103900                                  4  ...                       5   
103901                                  1  ...                       4   
103902                                  1  ...                       1   
103903                                  3  ...                       1   

        On-board service  Leg room service  Baggage handling  Checkin service  \
0                      4                 3                 4                4   
1                      1                 5                 3                1   
2                      4                 3                 4                4   
3                      2                 5                 3                1   
4                      3                 4                 4                3   
...                  ...               ...               ...              ...   
103899                 3                 1                 4                2   
103900                 5                 5                 5                5   
103901                 3                 2                 4                5   
103902                 4                 5                 1                5   
103903                 1                 1                 4                4   

        Inflight service  Cleanliness  Departure Delay in Minutes  \
0                      5            5                          25   
1                      4            1                           1   
2                      4            5                           0   
3                      4            2                          11   
4                      3            3                           0   
...                  ...          ...                         

De la pagina de Kaggle extraemos la siguiente información:

    Gender: Gender of the passengers (Female, Male)
    Customer Type: The customer type (Loyal customer, disloyal customer)
    Age: The actual age of the passengers
    Type of Travel: Purpose of the flight of the passengers (Personal Travel, Business Travel)
    Class: Travel class in the plane of the passengers (Business, Eco, Eco Plus)
    Flight distance: The flight distance of this journey
    Inflight wifi service: Satisfaction level of the inflight wifi service (0:Not Applicable;1-5)
    Departure/Arrival time convenient: Satisfaction level of Departure/Arrival time convenient
    Ease of Online booking: Satisfaction level of online booking
    Gate location: Satisfaction level of Gate location
    Food and drink: Satisfaction level of Food and drink
    Online boarding: Satisfaction level of online boarding
    Seat comfort: Satisfaction level of Seat comfort
    Inflight entertainment: Satisfaction level of inflight entertainment
    On-board service: Satisfaction level of On-board service
    Leg room service: Satisfaction level of Leg room service
    Baggage handling: Satisfaction level of baggage handling
    Check-in service: Satisfaction level of Check-in service
    Inflight service: Satisfaction level of inflight service
    Cleanliness: Satisfaction level of Cleanliness
    Departure Delay in Minutes: Minutes delayed when departure
    Arrival Delay in Minutes: Minutes delayed when Arrival
    Satisfaction: Airline satisfaction level(Satisfaction, neutral or dissatisfaction)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

Empezamos realizando una exploración inicial de los datos con un info y un describe, para ver nulos y en datos numericos ver promedios, maximos, minimos, etc.

In [5]:
df.describe()

Unnamed: 0             id            Age  Flight Distance  \
count  103904.000000  103904.000000  103904.000000    103904.000000   
mean    51951.500000   64924.210502      39.379706      1189.448375   
std     29994.645522   37463.812252      15.114964       997.147281   
min         0.000000       1.000000       7.000000        31.000000   
25%     25975.750000   32533.750000      27.000000       414.000000   
50%     51951.500000   64856.500000      40.000000       843.000000   
75%     77927.250000   97368.250000      51.000000      1743.000000   
max    103903.000000  129880.000000      85.000000      4983.000000   

       Inflight wifi service  Departure/Arrival time convenient  \
count          103904.000000                      103904.000000   
mean                2.729683                           3.060296   
std                 1.327829                           1.525075   
min                 0.000000                           0.000000   
25%                 2.000000                           2.000000   
50%                 3.000000                           3.000000   
75%                 4.000000                           4.000000   
max                 5.000000                           5.000000   

       Ease of Online booking  Gate location  Food and drink  Online boarding  \
count           103904.000000  103904.000000   103904.000000    103904.000000   
mean                 2.756901       2.976883        3.202129         3.250375   
std                  1.398929       1.277621        1.329533         1.349509   
min                  0.000000       0.000000        0.000000         0.000000   
25%                  2.000000       2.000000        2.000000         2.000000   
50%                  3.000000       3.000000        3.000000         3.000000   
75%                  4.000000       4.000000        4.000000         4.000000   
max                  5.000000       5.000000        5.000000         5.000000   

        Seat comfort  Inflight entertainment  On-board service  \
count  103904.000000           103904.000000     103904.000000   
mean        3.439396                3.358158          3.382363   
std         1.319088                1.332991          1.288354   
min         0.000000                0.000000          0.000000   
25%         2.000000                2.000000          2.000000   
50%         4.000000                4.000000          4.000000   
75%         5.000000                4.000000          4.000000   
max         5.000000                5.000000          5.000000   

       Leg room service  Baggage handling  Checkin service  Inflight service  \
count     103904.000000     103904.000000    103904.000000     103904.000000   
mean           3.351055          3.631833         3.304290          3.640428   
std            1.315605          1.180903         1.265396          1.175663   
min            0.000000          1.000000         0.000000          0.000000   
25%            2.000000          3.000000         3.000000          3.000000   
50%            4.000000          4.000000         3.000000          4.000000   
75%            4.000000          5.000000         4.000000          5.000000   
max            5.000000          5.000000         5.000000          5.000000   

         Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  
count  103904.000000               103904.000000             103594.000000  
mean        3.286351                   14.815618                 15.178678  
std         1.312273                   38.230901                 38.698682  
min         0.000000                    0.000000                  0.000000  
25%         2.000000                    0.000000                  0.000000  
50%         3.000000                    0.000000                  0.000000  
75%         4.000000                   12.000000                 13.000000  
max         5.000000                 1592.000000               1584.000000

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          690 non-null    object 
 1   Age             690 non-null    object 
 2   Debt            690 non-null    float64
 3   Married         690 non-null    object 
 4   BankCustomer    690 non-null    object 
 5   EducationLevel  690 non-null    object 
 6   Ethnicity       690 non-null    object 
 7   YearsEmployed   690 non-null    float64
 8   PriorDefault    690 non-null    object 
 9   Employed        690 non-null    object 
 10  CreditScore     690 non-null    int64  
 11  DriversLicense  690 non-null    object 
 12  Citizen         690 non-null    object 
 13  ZipCode         690 non-null    object 
 14  Income          690 non-null    int64  
 15  ApprovalStatus  690 non-null    object 
dtypes: float64(2), int64(2), object(12)
memory usage: 86.4+ KB


In [9]:
print("Categoricas: ", df.select_dtypes(include=['object']).columns)
for c in df:
    lista = df[c].unique()
    if df[c].dtype == 'object':
        print(f"{c}:", lista)
print("Numericas: ", df.select_dtypes(include=['int64','float64']).columns)
for c in df:
    lista = df[c].unique()
    if df[c].dtype != 'object':
        print(f"{c}:", len(lista))

Categoricas:  Index(['Gender', 'Customer Type', 'Type of Travel', 'Class', 'satisfaction'], dtype='object')
Gender: ['Male' 'Female']
Customer Type: ['Loyal Customer' 'disloyal Customer']
Type of Travel: ['Personal Travel' 'Business travel']
Class: ['Eco Plus' 'Business' 'Eco']
satisfaction: ['neutral or dissatisfied' 'satisfied']
Numericas:  Index(['Unnamed: 0', 'id', 'Age', 'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes'],
      dtype='object')
Unnamed: 0: 103904
id: 103904
Age: 75
Flight Distance: 3802
Inflight wifi service: 6
Departure/Arrival time convenient: 6
Ease of Online booking: 6
Gate location: 6
Food and drink: 

Observamos como varias de las columnas numericas tienen entre 5 y 6 valores distintos unicamente, aunque tenemos 103904 instancias, podriamos asumir que son los puntajes asignados, pero igual procedemos a mostrar dichos valores.

In [10]:
df[['Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness']].head(15)

Inflight wifi service  Departure/Arrival time convenient  \
0                       3                                  4   
1                       3                                  2   
2                       2                                  2   
3                       2                                  5   
4                       3                                  3   
5                       3                                  4   
6                       2                                  4   
7                       4                                  3   
8                       1                                  2   
9                       3                                  3   
10                      4                                  5   
11                      2                                  4   
12                      1                                  4   
13                      4                                  2   
14                      3                                  2   

    Ease of Online booking  Gate location  Food and drink  Online boarding  \
0                        3              1               5                3   
1                        3              3               1                3   
2                        2              2               5                5   
3                        5              5               2                2   
4                        3              3               4                5   
5                        2              1               1                2   
6                        2              3               2                2   
7                        4              4               5                5   
8                        2              2               4                3   
9                        3              4               2                3   
10                       5              4               2                5   
11                       2              2               1                2   
12                       4              4               1                1   
13                       4              3               4                4   
14                       3              2               2                3   

    Seat comfort  Inflight entertainment  On-board service  Leg room service  \
0              5                       5                 4                 3   
1              1                       1                 1                 5   
2              5                       5                 4                 3   
3              2                       2                 2                 5   
4              5                       3                 3                 4   
5              1                       1                 3                 4   
6              2                       2                 3                 3   
7              5                       5                 5                 5   
8              3                       1                 1                 2   
9              3                       2                 2                 3   
10             2                       2                 3                 3   
11             1                       1                 1                 2   
12             1                       1                 1                 1   
13             4                       4                 4                 5   
14             2                       2                 4                 3   

    Baggage handling  Checkin service  Inflight service  Cleanliness  
0                  4                4                 5            5  
1                  3                1                 4            1  
2                  4                4                 4            5  
3                  3                1                 4            2  
4                  4                3                 3            3  
5                  4   

Considerando lo anterior, vamos a realizar una primera prueba eliminando las columnas categoricas y hacemos una prueba inicial para obtener un Score desde el cual vamos a partir a mejorar. Observamos tambien que el tarjet se encuentra categorico con 'neutral or dissatisfied' y 'satisfied', razon por la cual utilizamos el LabelEncoder.

In [38]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.pipeline import Pipeline

from sklearn.preprocessing import LabelEncoder

df_clean = df.dropna(axis=0, how='any')
print("Conteo posterior a la eliminación de nulos: ", df_clean.shape)
label_encoder = LabelEncoder()  # OJO! LabelEncoder para la target
df_clean["satisfaction"] = label_encoder.fit_transform(df_clean["satisfaction"])
data = df_clean.select_dtypes(include=[np.number])

X, y = data.drop(columns=["satisfaction"]), data["satisfaction"]

Conteo posterior a la eliminación de nulos:  (103594, 25)


In [43]:
model1 = RandomForestClassifier()
pipeline1 = Pipeline(steps=[('m', model1)])
# Definimos la evaluación del modelo y lo aplicamos
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=96)
scores1 = cross_val_score(pipeline1, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy eliminando los nulos: %.3f (%.3f) RandomForestClassifier' % (scores1.mean(), scores1.std()))

Mean Accuracy eliminando los nulos: 0.949 (0.002) RandomForestClassifier


In [40]:
model2 = DecisionTreeClassifier()
pipeline2 = Pipeline(steps=[('m', model2)])
# Definimos la evaluación del modelo y lo aplicamos
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=96)
scores2 = cross_val_score(pipeline2, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy eliminando los nulos: %.3f (%.3f) DecisionTreeClassifier' % (scores2.mean(), scores2.std()))

Mean Accuracy eliminando los nulos: 0.921 (0.002) DecisionTreeClassifier


In [41]:
model3 = KNeighborsClassifier()
pipeline3 = Pipeline(steps=[('m', model3)])
# Definimos la evaluación del modelo y lo aplicamos
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=96)
scores3 = cross_val_score(pipeline3, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy eliminando los nulos: %.3f (%.3f) KNeighborsClassifier' % (scores3.mean(), scores3.std()))

Mean Accuracy eliminando los nulos: 0.600 (0.005) KNeighborsClassifier


In [42]:
model4 = LogisticRegression()
pipeline4 = Pipeline(steps=[('m', model4)])
# Definimos la evaluación del modelo y lo aplicamos
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=96)
scores4 = cross_val_score(pipeline4, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
print('Mean Accuracy eliminando los nulos: %.3f (%.3f) LogisticRegression' % (scores4.mean(), scores4.std()))

Mean Accuracy eliminando los nulos: 0.664 (0.013) LogisticRegression


Observamos que el mejor Score lo tenemos con el algoritmo RandomForestClassifier, llegando a 0.949. Empezamos a codificar las variables categoricas y demas para ver si podemos mejorar este puntaje. Para ello nuevamente comparamos la cantidad de categorias por cada columna para empezar a validar que tecnicas de preparacion vamos a usar.

In [44]:
for c in df:
    lista = df[c].unique()
    if lista.size < 20:
        print(f"{c}:", lista)
print(df.info())

Gender: ['Male' 'Female']
Customer Type: ['Loyal Customer' 'disloyal Customer']
Type of Travel: ['Personal Travel' 'Business travel']
Class: ['Eco Plus' 'Business' 'Eco']
Inflight wifi service: [3 2 4 1 5 0]
Departure/Arrival time convenient: [4 2 5 3 1 0]
Ease of Online booking: [3 2 5 4 1 0]
Gate location: [1 3 2 5 4 0]
Food and drink: [5 1 2 4 3 0]
Online boarding: [3 5 2 1 4 0]
Seat comfort: [5 1 2 3 4 0]
Inflight entertainment: [5 1 2 3 4 0]
On-board service: [4 1 2 3 5 0]
Leg room service: [3 5 4 2 1 0]
Baggage handling: [4 3 5 1 2]
Checkin service: [4 1 3 5 2 0]
Inflight service: [5 4 3 1 2 0]
Cleanliness: [5 1 2 3 4 0]
satisfaction: ['neutral or dissatisfied' 'satisfied']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   

Comenzamos revisando los nulos de la columna Arrival Delay in Minutes 

In [52]:
for i in df:
    # count number of rows with missing values
    n_miss = df[i].isna().sum()
    if n_miss > 0:
        perc = n_miss / df.shape[0] * 100
        print('> %s, Missing: %d (%.1f%%)' % (i, n_miss, perc))
df[df.isna().any(axis=1)]

> Arrival Delay in Minutes, Missing: 310 (0.3%)


Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
213            213   49608  Female     Loyal Customer   38  Business travel   
1124          1124   73442    Male     Loyal Customer   53  Personal Travel   
1529          1529   71178    Male     Loyal Customer   39  Business travel   
2004          2004   72940  Female  disloyal Customer   26  Business travel   
2108          2108  116374  Female     Loyal Customer   24  Personal Travel   
...            ...     ...     ...                ...  ...              ...   
102067      102067   36729    Male     Loyal Customer   49  Personal Travel   
102384      102384   71241    Male     Loyal Customer   58  Business travel   
102552      102552   27684  Female  disloyal Customer   29  Business travel   
102960      102960   36787    Male     Loyal Customer   58  Business travel   
103540      103540   45022  Female     Loyal Customer   33  Personal Travel   

           Class  Flight Distance  Inflight wifi service  \
213          Eco              109                      5   
1124         Eco             1012                      3   
1529    Business              733                      2   
2004    Business             1035                      3   
2108         Eco              417                      2   
...          ...              ...                    ...   
102067  Eco Plus             1249                      2   
102384       Eco              733                      3   
102552       Eco             1107                      2   
102960       Eco             1088                      4   
103540       Eco              359                      4   

        Departure/Arrival time convenient  ...  Inflight entertainment  \
213                                     3  ...                       5   
1124                                    2  ...                       4   
1529                                    5  ...                       2   
2004                                    3  ...                       2   
2108                                    1  ...                       5   
...                                   ...  ...                     ...   
102067                                  5  ...                       3   
102384                                  3  ...                       3   
102552                                  1  ...                       5   
102960                                  4  ...                       5   
103540                                  4  ...                       4   

        On-board service  Leg room service  Baggage handling  Checkin service  \
213                    5                 2                 4                1   
1124                   4                 4                 4                3   
1529                   2                 2                 2                2   
2004                   3                 3                 4                5   
2108                   1                 4                 2                1   
...                  ...               ...               ...              ...   
102067                 4                 5                 4                3   
102384                 3                 1                 2                4   
102552                 4                 1                 5                5   
102960                 1                 5                 5                5   
103540                 3                 2                 5                3   

        Inflight service  Cleanliness  Departure Delay in Minutes  \
213                    1            5                          31   
1124                   3            4                          38   
1529                   2            3                          11   
2004                   5            2                          41   
2108                   2            5                           1   
...                  ...          ...                         

Considerando la naturaleza de la colummna y evidenciando que son solamente el 0.3% de los datos, decidimos borrar esas instancias para evitar problemas con el modelo.

In [54]:
df.dropna(inplace=True)

Empezamos el tratamiento separando las columnas para realizar tratamientos independientes.

In [67]:
df_trated = df.copy()
y = df_trated['satisfaction']
df_trated.drop(['satisfaction'], axis=1, inplace=True)
categorical_ix = df_trated.select_dtypes(include='object').columns 
numerical_ix = df_trated.select_dtypes(include=['float64','int']).columns

De acuerdo a la info hecho mas arriba observamos que para la mayoria de caracteristicas categoricas contamos con unicamente dos valores, pero para la caracteristica 'Class' tenemos 3 categorias, las cuales pueden ser ordenadas de forma nominal, razon por la cual procedemos a codificarlas de esta manera.

In [69]:
from sklearn.preprocessing import OneHotEncoder, MinMaxScaler, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from numpy import mean, std, delete

cat_without_target = delete(categorical_ix,-1) # Eliminamos la caracteristica Class de la lista de categoricas, esto para poder realizar un procesamiento especial con ella

preprocessor = ColumnTransformer(
    transformers=[
        ('ord_class', OrdinalEncoder(categories=[['Eco','Eco Plus', 'Business']]), ['Class']),
        ('ordinal', OrdinalEncoder(), cat_without_target),
    ],
    remainder='passthrough'
)

# define the data preparation and modeling pipeline
pipeline = Pipeline(steps=[('prep',preprocessor), ('m', model1)])

# evaluate the pipeline
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
n_scores = cross_val_score(pipeline, df_trated, y, scoring='accuracy', cv=cv, n_jobs=-1)
# report pipeline performance
print('Accuracy: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))

Accuracy: 0.964 (0.001)


Con las tecnicas aplicadas hasta el momento podemos ver una mejora en el accuracy, pero igualmente vamos tambien a revisar tecnicas sobre las variables numericas y posteriormente realizar una seleccion de caracteristicas.